cranelift/isle/isle/src/sema.rs

```diff
@@ -1543,36 +1543,51 @@ impl TermEnv {
                 Some((Pattern::BindPattern(ty, id, Box::new(subpat)), ty))
             }
             &ast::Pattern::Var { ref var, pos } => {
-                // Look up the variable; it must already have been bound.
+                // Look up the variable; if it has already been bound,
+                // then this becomes a `Var` node (which matches the
+                // existing bound value), otherwise it becomes a
+                // `BindPattern` with a wildcard subpattern to capture
+                // at this location.
                 let name = tyenv.intern_mut(var);
-                let bv = match bindings.vars.iter().rev().find(|bv| bv.name == name) {
+                match bindings.vars.iter().rev().find(|bv| bv.name == name) {
```

Comment:
Can we keep a set/map of already-bound variables? Doing a linear search here seems suboptimal.

Reply:
Ah, I think I originally wrote it this way for two reasons:

- It allows us to push/pop bindings when we enter and leave a lexical sub-scope (a `let`) -- we need to record the nesting order somehow, so a map to accelerate this would have to be maintained using that linear order on the side;
- It makes it easier to refactor to allow for shadowing later (which is a feature I think we might want).

In practice with a handful of bindings it shouldn't really be a perf concern, I think...

Reply:
(If you still want us to do something better here I'm happy to think about it more, though!)

Reply:
Eh, if it isn't a quick fix then lets leave it and deal with it if/when it actually becomes a problem